prow/jenkins/jenkins.go

```diff
@@ -401,14 +424,147 @@ func getJobName(spec *prowapi.ProwJobSpec) string {
 	return spec.Job
 }
 
-// getRequestPath builds an approriate path to use for this Jenkins Job.
-func getRequestPath(spec *prowapi.ProwJobSpec) string {
+// getJobInfoPath builds an approriate path to use for this Jenkins Job to get the job information
+func getJobInfoPath(spec *prowapi.ProwJobSpec) string {
+	jenkinsJobName := getJobName(spec)
+	jenkinsPath := fmt.Sprintf("/job/%s/api/json", jenkinsJobName)
+
+	return jenkinsPath
+}
+
+// getBuildPath builds a path to trigger a regular build for this job
+func getBuildPath(spec *prowapi.ProwJobSpec) string {
+	jenkinsJobName := getJobName(spec)
+	jenkinsPath := fmt.Sprintf("/job/%s/build", jenkinsJobName)
+
+	return jenkinsPath
+}
+
+// getBuildWithParametersPath builds a path to trigger a build with parameters for this job
+func getBuildWithParametersPath(spec *prowapi.ProwJobSpec) string {
 	jenkinsJobName := getJobName(spec)
 	jenkinsPath := fmt.Sprintf("/job/%s/buildWithParameters", jenkinsJobName)
 
 	return jenkinsPath
 }
 
+// GetJobInfo retrieves Jenkins job information
+func (c *Client) GetJobInfo(spec *prowapi.ProwJobSpec) (*JobInfo, error) {
+	path := getJobInfoPath(spec)
+	c.logger.Debugf("getJobInfoPath: %s", path)
+
+	data, err := c.Get(path)
+
+	if err != nil {
+		c.logger.Errorf("Failed to get job info: %v", err)
+		return nil, err
+	}
+
+	var jobInfo JobInfo
+
+	if err := json.Unmarshal(data, &jobInfo); err != nil {
+		return nil, fmt.Errorf("Cannot unmarshal job info from API: %v", err)
+	}
+
+	c.logger.Tracef("JobInfo: %+v", jobInfo)
+
+	return &jobInfo, nil
+}
+
+// JobParameterized tells us if the Jenkins job for this ProwJob is parameterized
+func (c *Client) JobParameterized(jobInfo *JobInfo) bool {
+	for _, prop := range jobInfo.Property {
+		if prop.ParameterDefinitions != nil && len(prop.ParameterDefinitions) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// EnsureBuildableJob attempts to detect a job that hasn't yet ran and populated
+// its parameters. If detected, it tries to run a build until the job parameters
+// are processed, then it aborts the build.
+func (c *Client) EnsureBuildableJob(spec *prowapi.ProwJobSpec) error {
+	jobInfo, err := c.GetJobInfo(spec)
```

Comment:
Oh yeah, about this.

It's possible that this will fail with `404` on a new PR because Github hook hasn't yet fired or hasn't been processed yet and Jenkins doesn't yet know about this job. I thought about putting a 1-time retry here (say, wait 10 sec, try again, if 404 again the fail), thoughts?